internal/topology/check/compatibility.go

```diff
@@ -176,10 +177,153 @@ func LocalObjectTemplateIsValid(in *clusterv1.LocalObjectTemplate, namespace str
+// It checks that:
+// 1) Infrastructure Templates are compatible.
+// 2) ControlPlane Templates are compatible.
+// 3) ControlPlane Machine Infrastructure Templates are compatible.
+// 4) MachineDeploymentClasses have not been deleted and are compatible.
+func ClusterClassesAreCompatible(current, desired *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if current == nil {
+		return nil
+	}
+
+	// Validate InfrastructureClusterTemplate changes desired a compatible way.
+	allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(current.Spec.Infrastructure, desired.Spec.Infrastructure)...)
+
+	// Validate control plane changes desired a compatible way.
+	allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(current.Spec.ControlPlane.LocalObjectTemplate, desired.Spec.ControlPlane.LocalObjectTemplate)...)
+	if desired.Spec.ControlPlane.MachineInfrastructure != nil && current.Spec.ControlPlane.MachineInfrastructure != nil {
+		allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(*current.Spec.ControlPlane.MachineInfrastructure, *desired.Spec.ControlPlane.MachineInfrastructure)...)
+	}
+
+	// Validate changes to MachineDeployments.
+	allErrs = append(allErrs, MachineDeploymentClassesAreCompatible(current, desired)...)
+
+	return allErrs
+}
+
+// MachineDeploymentClassesAreCompatible checks if each MachineDeploymentClass in the new ClusterClass is a compatible change from the previous ClusterClass.
+// It checks if:
+// 1) Any MachineDeploymentClass has been removed.
+// 2) If the MachineDeploymentClass.Template.Infrastructure reference has changed its Group or Kind.
+func MachineDeploymentClassesAreCompatible(current, desired *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Ensure no MachineDeployment class was removed.
+	classes := classNamesFromWorkerClass(desired.Spec.Workers)
+	for _, oldClass := range current.Spec.Workers.MachineDeployments {
+		if !classes.Has(oldClass.Class) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments"),
+					desired.Spec.Workers.MachineDeployments,
+					fmt.Sprintf("The %q MachineDeployment class can't be removed.", oldClass.Class),
+				),
+			)
+		}
+	}
+
+	// Ensure previous MachineDeployment class was modified in a compatible way.
+	for _, class := range desired.Spec.Workers.MachineDeployments {
+		for _, oldClass := range current.Spec.Workers.MachineDeployments {
+			if class.Class == oldClass.Class {
+				// NOTE: class.Template.Metadata and class.Template.Bootstrap are allowed to change;
+				// class.Template.Bootstrap are ensured syntactically correct by LocalObjectTemplateIsValid.
+				// Validates class.Template.Infrastructure template changes in a compatible way
+				allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(oldClass.Template.Infrastructure, class.Template.Infrastructure)...)
```

Comment:
I wonder if we should pass in the relative fldPath (including index here) like in LocalObjectTemplateIsValid below. I saw this in the upstream CRD validation all over and I think it makes a lot of sense (and afaik we tried to do this consistently in our other webhooks)

Right now the user would get an error that `.metadata.apiVersion` is invalid (which looks like the top-level ClusterClass metadata), although it actually is `.spec.workers.machineDeployments[i].template.infrastructure.ref.metadata.apiVersion`

Reply:
As a general rule. I think every (field) validation func which doesn't operate on the root resource should take a fldPath to provide errors with correct field paths.

Reply:
I'm not sure. The user will update a ClusterClass and will get an error that `.metadata.apiVersion` is invalid. As Kubernetes is usually reporting the full path I'm not sure if it's easier to troubleshoot.

Reply:
@fabriziopandini Do you have an opinion on that?

Reply:
+1 to get field paths

Reply:
Sorry - I was referring to this comment above:

> I usually would prefer using the index (and not the class or topology name) as otherwise it looks like a map[string]something.

Current plan is to add the index and path to these.